target/linux/ath79/generic/base-files/lib/upgrade/openmesh.sh

```diff
@@ -0,0 +1,114 @@
+# The U-Boot loader of the OpenMesh devices requires image sizes and
+# checksums to be provided in the U-Boot environment.
+# The OpenMesh devices come with 2 main partitions - while one is active
+# sysupgrade will flash the other. The boot order is changed to boot the
+# newly flashed partition. If the new partition can't be booted due to
+# upgrade failures the previously used partition is loaded.
+
+trim()
+{
+	echo $1
+}
+
+cfg_value_get()
+{
+	local cfg=$1 cfg_opt
+	local section=$2 our_section=0
+	local param=$3 our_param=
+
+	for cfg_opt in $cfg
+		do
+			[ "$cfg_opt" = "[$section]" ] && our_section=1 && continue
+			[ "$our_section" = "1" ] || continue
+
+			our_param=$(echo ${cfg_opt%%=*})
+			[ "$param" = "$our_param" ] && echo ${cfg_opt##*=} && break
+		done
+}
+
+platform_do_upgrade_openmesh()
+{
+	local img_path="$1"
+	local restore_backup
+
+	local setenv_script="/tmp/fw_env_upgrade"
+
+	local inactive_mtd="$(find_mtd_index $PART_NAME)"
+	local inactive_offset="$(cat /sys/class/mtd/mtd${inactive_mtd}/offset)"
+	local total_size="$(cat /sys/class/mtd/mtd${inactive_mtd}/size)"
+	local flash_start_mem=0x9f000000
+	local data_offset=$((64 * 1024))
+
+	# detect to which flash region the new image is written to.
+	#
+	# 1. check what is the mtd index for the first flash region on this
+	#    device
+	# 2. check if the target partition ("inactive") has the mtd index of
+	#    the first flash region
+	#
+	#    - when it is: the new bootseq will be 1,2 and the first region is
+	#      modified
+	#    - when it isnt: bootseq will be 2,1 and the second region is
+	#      modified
+	#
+	# The detection has to be done via the hardcoded mtd partition because
+	# the current boot might be done with the fallback region. Let us
+	# assume that the current bootseq is 1,2. The bootloader detected that
+	# the image in flash region 1 is corrupt and thus switches to flash
+	# region 2. The bootseq in the u-boot-env is now still the same and
+	# the sysupgrade code can now only rely on the actual mtd indexes and
+	# not the bootseq variable to detect the currently booted flash
+	# region/image.
+	#
+	# In the above example, an implementation which uses bootseq ("1,2") to
+	# detect the currently booted image would assume that region 1 is booted
+	# and then overwrite the variables for the wrong flash region (aka the
+	# one which isn't modified). This could result in a device which doesn't
+	# boot anymore to Linux until it was reflashed with ap51-flash.
+	local next_boot_part="1"
+	local primary_kernel_mtd="3"
+	[ "$inactive_mtd" = "$primary_kernel_mtd" ] || next_boot_part="2"
+
+	local cfg_size=$(dd if="$img_path" bs=2 skip=35 count=4 2>/dev/null)
+	local cfg_content=$(dd if="$img_path" bs=1 skip=$data_offset count=$((0x$cfg_size)) 2>/dev/null)
```

Comment:
That's also a candidate for swapping bs and count (needs additional skip-bytes then, of course).

Reply:
Please don't swap bs and count. Will end up easily with copy problems on busybox with MIPS. I've tried it for the sysupgrade-tar on PA300 and ended up with half "dd"ed images. Most likely memory problems which caused the dd to stop and thus the tar to have write errors to the pipe.

Reply:
We are using skip_bytes and seek_bytes in various places, e.g. for caldata extraction (ath9k/ath10k). Why would this be different here?

I agree that dd-ing an entire kernel/rootfs in one chunk is not really a good idea. But doing so with bs=1 isn't desirable either. And talking about cfg_content here, this is reading (not writing) and probably smaller anyway.

But anyway, this is a separate discussion, I would definitely not touch bs/count for this commit/PR.

Reply:
Ok, just checked the source code and you seem to be right. My statement was wrong that this isn't compiled in by default in openwrt.

Reply:
And since we assume blocksize-multiples for kernel (and rootfs as well?) now anyway, we could relatively simple work with bs=1k or similar there (even something like bs=64 would probably be an enormous speed improvement without jeopardizing the whole process too much).

Reply:
Just to repeat the current results with bs=1024 from the other comment thread: Decreases the write time of the default image (without feeds) on an OM2P-HS v3 from roughly 120s to 30s. The erase + some overhead add in both cases roughly another 50s

Reply:
Thanks for testing, that's really a speedup, isn't it ;-)

But note that even in case I convinced you, this (bs vs. count) should be a separate patch, so we can revert/alter it easily later.

Reply:
Problem is only that it doesn't work for the rootfs because the rootfs is (64k or 256k) * number of blocks + 4 bytes. Here we would need the count_bytes or an ugly hack to append the last "block".

Reply:
Theoretically, one could use bs=4 and would already be faster, but that's not really less ugly.

I wonder whether we can just get around that with `conv=sync` or similar. A few additional zeros should not hurt, and make everything easier.

I've never dealt with count_bytes so far.

Reply:
wouldn't be a couple of zeros because the metadata follows after the rootfs

Reply:
I have now added a minor change (with large boilerplate) which just rounds up the number of 1024 byte blocks for rootfs - which means that it write the "End of filesystem" marker to flash + in some circumstances bogus sysupgrade metadata.

These "circumstances" would be when no config would be saved during a sysupgrade.